### PR TITLE
breaking: add support for ressources with relative paths

### LINKED ISSUE #26725: Cypress CT does not support ressources fetched from relative paths
### Current behavior

Any relative request from inside a component test will include `/__cypress/iframe` and will therefore be mapped to `/__cypress/src/index.html` and forwarded to the ct-dev-server. A header is provided including the original request path, however it would be quite hacky to use this to parse the relative path, as it would need to be matched against the test-file it was last called with.

### Desired behavior

Cypress should be able to distinguish between the initial request used to kickstart the test-runner that needs to be mapped to the index.html and subsequent relative requests, that need to be forwarded to the dev-server as relative requests.

i.e. a request to `./data/image.png` in a test with absolute path `c:/project/cypress/component/test.cy.jsx` currently resolves to `/__cypress/iframe/c:/project/cypress/component/data/image.png` but should resolve to `/__cypress/src/data/image.png`

### Test code to reproduce

```javascript
it('image with relative path should load', () => {
  cy.mount(<img src="./path/to/image.png" />)
  cy.get('img').should('be.visible').then(($img) => {
    expect($img[0].naturalWidth).to.be.greaterThan(0);
  })
})
```
I have a failing testcase in my custom dev-server here:
https://github.com/fochlac/cypress-ct-custom-devserver/blob/main/cypress/component/Test.cy.tsx#L13-L27

However I have also reproduced the issue using the webpack dev server in an private repository. Since the offending code is in the iframe controller this should also affect the vite-devserver:
https://github.com/cypress-io/cypress/blob/7630caed820c685295ff2fb52c98455ab0cf7bf6/packages/server/lib/controllers/iframes.ts#L44C1-L62

A workaround would be to use the `__cypress_spec_path `-header and match it against subsequent requests to properly map relative requests to the output folder, but that is super hacky.

### Cypress Version

12.x

### Node version

16.16

### Operating System

windows / unix

### Debug Logs

```shell
GET /__cypress/iframes/C:/gitHub/cypress-ct-custom-devserver/cypress/component/Test.cy.tsx 304 2.459 ms - 
GET /__cypress/iframes/C:/gitHub/cypress-ct-custom-devserver/cypress/component/data/image.jpg 304 2.484 ms
```

### Other

_No response_

### PR DESCRIPTION
**Pull request by fochlac**
- Closes [26725](https://github.com/cypress-io/cypress/issues/26725)

### Steps to test
create a cypress test that mounts a static ressource via a relative path, i.e. `` and check whether the request succeds

### How has the user experience changed?
It is now possible to distinguish the initial request for the index.html of the CT from relative asset requests (fonts/images/etc). This allows routing those requests to dev-server while retaining the relative paths.
Example: if a component test includes an image defined as follows: `` the load image request will be forwarded to the url `dev-server:port/__cypress/src/path/to/img.png` instead of `dev-server:port/__cypress/src/index.html`.

### PR Tasks

- [x] Have tests been added/updated?
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

File: system-tests/lib/dep-installer/index.ts [hunk 1]
```diff
@@ -99,7 +99,7 @@ function getRelativePathToProjectDir (projectDir: string) {
 async function restoreLockFileRelativePaths (opts: { projectDir: string, lockFilePath: string, relativePathToMonorepoRoot: string }) {
   const relativePathToProjectDir = getRelativePathToProjectDir(opts.projectDir)
   const lockFileContents = (await fs.readFile(opts.lockFilePath, 'utf8'))
-  .replaceAll(opts.relativePathToMonorepoRoot, relativePathToProjectDir)
+  .replaceAll(opts.relativePathToMonorepoRoot.replace(/\\+/g, '/'), relativePathToProjectDir.replace(/\\+/g, '/'))
```
Comment (lmiller1990) [inline system-tests/lib/dep-installer/index.ts:102]: Why was this change necessary?

File: system-tests/test/webpack_dev_server_spec.ts [hunk 2]
```diff
@@ -45,6 +45,24 @@ describe('@cypress/webpack-dev-server', function () {
     },
   })
 
+  systemTests.it('successfully loads and runs spec to check if relative urls are working', {
```
Comment (lmiller1990) [inline system-tests/test/webpack_dev_server_spec.ts]: Test looks good, but I think we can do one better by writing the test here: https://github.com/cypress-io/cypress/blob/develop/npm/webpack-dev-server/cypress/e2e/webpack-dev-server.cy.ts

This will give you "Cypress in Cypress" so you can actually assert the test runs, assets load, etc. What do you think? You can still use the same test project you created.

File: npm/webpack-dev-server/src/loader.ts [hunk 3]
```diff
@@ -18,7 +18,7 @@ const makeImport = (file: Cypress.Cypress['spec'], filename: string, chunkName:
   const magicComments = chunkName ? `/* webpackChunkName: "${chunkName}" */` : ''
 
   return `"${filename}": {
-    shouldLoad: () => decodeURI(document.location.pathname).includes("${file.absolute}"),
+    shouldLoad: () => (new URL(document.location)).searchParams.get("specPath") === "${file.absolute}",
```
Comment (lmiller1990) [inline npm/webpack-dev-server/src/loader.ts:21]: I think this should be fine, let's see how CI goes.

File: system-tests/lib/dep-installer/index.ts (hunk 1 above)
Comment (fochlac) [reply system-tests/lib/dep-installer/index.ts:102]: This was neccessary to be able to run a project with a linked project in the yarn lockfile in windows. yarn always uses `/` to resolve paths, the path-seperators used by path however change depending on the os.

File: npm/webpack-dev-server/src/loader.ts (hunk 3 above)
Comment (fochlac) [reply npm/webpack-dev-server/src/loader.ts:21]: It works, however, this makes it a breaking change I think. This will become a problem if the cypress-version doesn't match the webpack-dev-server version, and I don't know if I can assume this to always be the case.
I considered supporting both variants to be backward compatible, but was not sure if its actually needed.

File: system-tests/test/webpack_dev_server_spec.ts (hunk 2 above)
Comment (fochlac) [reply system-tests/test/webpack_dev_server_spec.ts]: I thought the systemTests do kind of the same, but tbh. the testing structure is kind of complex and somewhat intransparent. ^^ I'll try to move the test over :)

File: system-tests/lib/dep-installer/index.ts (hunk 1 above)
Comment (lmiller1990) [reply system-tests/lib/dep-installer/index.ts:102]: Right, I see. We only run the system-tests on linux on CI normally, and most developers are on linux/mac, so I suppose that's why no-one had encountered this before.

File: system-tests/test/webpack_dev_server_spec.ts (hunk 2 above)
Comment (lmiller1990) [reply system-tests/test/webpack_dev_server_spec.ts]: Kind of - the system tests are basically testing "run" mode - like `npx cypress run`. The other tests I linked (which we call "Cypress in Cypress") will let you test against a Cypress instance running in "open" mode, as in `npx cypress open`. You can also make assertions against the Cypress instance. Ultimately, either place is probably fine for the test - if it's too much effort to move it, I wouldn't worry, but it's nice to keep the webpack tests in the `npm/webpack-dev-server` directory where possible.

File: npm/webpack-dev-server/src/loader.ts (hunk 3 above)
Comment (lmiller1990) [reply npm/webpack-dev-server/src/loader.ts:21]: Why is this a breaking change? Can you share some examples of when `shouldLoad` will be return a different value after your change?

Making a breaking change is definitely not something we take lightly, I think we want to avoid this if possible. Otherwise, this change would need to wait for the next major - a lot of planning goes into these.

File: system-tests/test/webpack_dev_server_spec.ts (hunk 2 above)
Comment (fochlac) [reply system-tests/test/webpack_dev_server_spec.ts]: I already moved it. :)

File: npm/webpack-dev-server/src/loader.ts (hunk 3 above)
Comment (fochlac) [reply npm/webpack-dev-server/src/loader.ts:21]: If this change would be part of 12.13.1 and some user uses cypress@12.13.1 and @cypress/webpack-dev-server@12.13.0 (or the other way round) no tests would load.

Comment (lmiller1990) [reply npm/webpack-dev-server/src/loader.ts:21]: Right, gotcha. Actually this won't be a big problem, but it will be breaking.

We'd wait for Cypress 13, and update this table (from the npm module readme: https://www.npmjs.com/package/@cypress/webpack-dev-server)

image

Let me review and test this a bit more. This will need to be part of v13.

Comment (lmiller1990) [reply npm/webpack-dev-server/src/loader.ts:21]: I'd missed that `document.pathname` does NOT include search params. I understand the breaking change now.

**Review by lmiller1990 — COMMENTED** [review]
This looks pretty good, let me sync with the team on how to get this in v13.

**Comment by lmiller1990** [comment]
@fochlac I talked to the team, this will be able to go in Cypress 13. That is at least a month away, so this won't merge for a while, but I'll get some more reviewers on it for you, and we can communicate as Cypress 13 draws closer.

For reference and for other reviewers:

*What’s the breaking change that users would experience?*

If a user has installed `@cypress/webpack-dev-server` from npm and configured their own dev-server, then upgraded to a version of Cypress with this change, it would break (the spec would not load).

This workflow (install `@cypress/webpack-dev-server` manually, instead of using the one that we ship with the binary) is pretty uncommon, it’ll only impact a handful of power users. The change here is an implementation detail. It’s only breaking because the contract between our npm module `@cypress/webpack-dev-server` and the binary is not guaranteed if you install `@cypress/webpack-dev-server` on your own instead of the one we include in the binary.

*What docs would need to be changed?*

Nothing in our main docs, we’d just [update this table](https://github.com/cypress-io/cypress/tree/develop/npm/webpack-dev-server#compatibility). This is actually an implementation detail, not part of the public API.

**Comment by fochlac** [comment]
@lmiller1990 sounds great, waiting for Cypress 13 is no big deal. Thanks for summarizing the breaking changes. I guess I was too deep into the topic to see why it's not obvious. :)

**Comment by lmiller1990** [comment]
@fochlac thought about this one a bit more - what do you think about keeping the old conditional, and including the new one with an or statement? No breaking change that way - it'll minimize any risk, and we could even release it in a minor/patch.

**Comment by fochlac** [comment]
@lmiller1990 I thought about this as well, but I think it just covers a less common case.
Let's have a look at the options (considering this is released with 13.0.0):
1) @cypress/webpack-dev-server@12.13.0 and cypress@13.0.0 -> tests will not load, as the test-path is in the query params and dev-server checks pathname
2) @cypress/webpack-dev-server@13.0.0 and cypress@12.13.0 -> tests will not load, as test-path is in pathname but webpack checks query params

That change would prevent case 2, however I think that one is the one that is much less likely to occur, that's why I have not left in the alternative check. If you think it's worth it, I'm open to add it though.

**Comment by lmiller1990** [comment]
We've got a few weeks to think about it either way. The cleanest way is likely the breaking change as you are pointing out, but glad we considered both angles.

We could potentially add a check in Cypress 13 that looks for an older version of `@cypress/webpack-dev-server` and throws an error such as "You are on Cypress 13, but using an old version of `@cypress/webpack-dev-server`. Please update `@cypress/webpack-dev-server`` to version 4". 

It would be v4 since the current one is v3: https://www.npmjs.com/package/@cypress/webpack-dev-server - the Cypress major isn't tied to the `@cypress/webpack-dev-server` version.

Just to clarify, we ship the correct version of `@cypress/webpack-dev-server` in the binary as part of the `cypress` package - only users who install `@cypress/webpack-dev-server` manually (who might want to override the default behavior) will be impacted by this - anyone using the defaults will automatically get updated when they install Cypress 13, so this shouldn't be a problem for the majority of users.

**Comment by lmiller1990** [comment]
Hey @fochlac, we might want to rebase and retarget this to the v13 branch: https://github.com/cypress-io/cypress/pull/27040

File: npm/webpack-dev-server/src/loader.ts (hunk 3 above)
Comment (fochlac) [reply npm/webpack-dev-server/src/loader.ts:21]: I extended the table, should I also update the changelog?

**Comment by fochlac** [comment]
Thanks for the head's up. I rebased and retargeted the PR. I also adapted the table you mentioned. From your PoV what are the remaining tasks I need to do for this PR?

**Comment by lmiller1990** [comment]
I don't think any additional work is needed on your end. Let me run CI and see if this is still green.

**Comment by lmiller1990** [comment]
Cy 13 coming soon. We will get this ready go to.

File: npm/webpack-dev-server/README.md [hunk 4]
```diff
@@ -68,6 +68,7 @@ We then merge the sourced config with the user's webpack config, and layer on ou
 | --------------------------- | ------- |
 | <= v1                       | <= v9   |
 | >= v2                       | >= v10  |
+| >= v4                       | >= v13  |
```
Comment (lmiller1990) [inline npm/webpack-dev-server/README.md:71]: @fochlac hey! We are looking to merge this soon. Should this be v3?

Also can you rebase again? Thank you!

If you can't, let me know and I can try to pick it up.

File: npm/webpack-dev-server/src/loader.ts (hunk 3 above)
Comment (lmiller1990) [reply npm/webpack-dev-server/src/loader.ts:21]: The changelog for npm is automated, no problem here.

**Comment by lmiller1990** [comment]
I resolved the conflicts, let's get this green and ready to merge.

File: npm/webpack-dev-server/README.md (hunk 4 above)
Comment (fochlac) [reply npm/webpack-dev-server/README.md:71]: Hi, only saw your comment just now. I think the new version should be v4 as we are currently on v3, and for full compatibility we need dev-server >=v4 and cypress >=v13.
We could consider adapting this like this as well for more clarity:
| @cypress/webpack-dev-server | cypress |
| --------------------------- | ------- |
| <= v1                       | <= v9   |
| v2 - v3                      | v10 - v12  |
| >= v4                       | >= v13  |

**Comment by lmiller1990** [comment]
CI has some issues w/ third party PRs right now. I will try re-opening a new PR with the same commits: https://github.com/cypress-io/cypress/pull/27510

**Comment by lmiller1990** [comment]
Closing for https://github.com/cypress-io/cypress/pull/27510

Same code but this one cannot pass CI for unknown reasons
